src/metaopt/core/worker/experiment.py

```diff
@@ -283,6 +283,9 @@ def configure(self, config):
         if self._init_done:
             raise RuntimeError("Configuration is done; cannot reset an Experiment.")
 
+        # Sanitize and replace some sections with objects
+        candidate_algorithms = Experiment._sanitize_and_instantiate_config(config)
```

Comment:
Ok this is a little bit nasty, but I will explain:

**What we 'want'**:
1. Most recent metaopt configuration has priority over the existing in database. In addition, having the same experiment name, means that we run the same algorithms.
2. Configurations should match except allowed exceptions. Currently those are: `    non_forking_attrs = ('status', 'pool_size', 'max_trials')`
3. User can specify an algorithm to run with its defaults parameters just by specifying the algo's name. So a user will not be forced to write for instance:
`algorithms: random: {}` instead of `algorithms: random` in moptconfig. Note that in former case the value of algorithms is a dictionary, while in the latter case it is a string.
4. We would like to save in the database the complete specification of an algorithm, with all of its parameters specified. That's a measure in order to have consistency. Consider the case where the defaults values for an algorithm's parameters change.

**What's the problem and solution**:
Comparing the configuration, needs them to have a similar form. For our convenience, we can proactively instantiate algorithms from the new configuration (so that (3) can be a feature) and sanitize it by executing (code from `Experiment._sanitize_and_instantiate_config`):
```
        algorithms = PrimaryAlgo(space, config['algorithms'])
        # Sanitize 'algorithms' section in given configuration
        config['algorithms'] = algorithms.configuration
```
, as `BaseAlgorithm.configuration` always returns the complete specification.
So now we can compare them safely, because we didn't mess non-forking attributes (2) and we brought it to a consistent form (4). We are able to proactively instantiate the algorithms because of (1).

Reply:
Can you elaborate on:
> I would at least split that line and fetch candidate_algorithms from config after instantiation.

I did not quite get it.

Reply:
As I said, I understand the convenience, but thanks for the explanation anyway. :slightly_smiling_face: 

Is `Experiment._sanitize_and_instantiate_config()` expected to affect the state of the experiment object at some point? I know it does not in the current implementation but maybe you have some plan in mind where this would happen? If it's not the case, then I would make the function edit a copy of config and return the modified version, leaving the passed one intact. 

```python
instantiated_config = Experiment._sanitize_and_instantiate_config(config)
candidate_algorithms = instantiated_config['algorithms']
```

Or if we keep current `inplace` implementation, my two line split proposition would look like this.

```python
Experiment._sanitize_and_instantiate_config(config)
candidate_algorithms = config['algorithms']
```

I'd prefer the former one.

Reply:
@tsirif Can we settle on this one too?

Reply:
I don't think it can be changed this way. or I am not able to see the way. I am going to go through the code with a particular example in mind (I will remove some code to focus where it's necessary):

```python
    def configure(self, config):
        # ``config['algorithms'] == 'random'`` ; this was set from user's configuration
        candidate_algorithms = Experiment._sanitize_and_instantiate_config(config)
        # after this call variable config will be: ``config['algorithms'] == {'random': {}}``

        if self.status is None:  # Suppose that experiment with same name already exists
            is_new = True
        else:  # So we will be here
            # Now we have to check against what already exists in database
            # If were not to to `_sanitize_and_instantiate_config` as it is,
            # then ``config['algorithms'] == 'random'``.
            is_new = self._is_different_from(config)
            # ``is_new is True``, if ``config['algorithms'] == 'random'``.
            # But ``is_new is False``, if ``config['algorithms'] == {'random': {}}``
            # That is because in database, a complete (not partial) specification of the experiment
            # and its algorithms is being saved.
            if is_new:
                self._fork_config(config)

        for section, value in config.items():
            setattr(self, section, value)

        self._init_done = True
        # In any case, current config from user has the top precedence in defining
        # which algorithms are going to be run.
        self.algorithms = candidate_algorithms
        self.status = 'pending'
        # Get complete specification of experiment
        final_config = self.configuration

        # Push it to the database
        if is_new:
            self._db.write('experiments', final_config)
            self._id = final_config['_id']
        else:
            self._db.write('experiments', final_config, {'_id': self._id})
```

```python
    def _sanitize_and_instantiate_config(config):
        space = SpaceBuilder().build_from(config['metadata']['user_args'])
        if not space:
            raise ValueError("Parameter space is empty. There is nothing to optimize.")
        # Instantiate algorithms
        algorithms = PrimaryAlgo(space, config['algorithms'])
        # Get back a complete specification of algorithms configuration!
        # Basically this step converts, ``config['algorithms'] == 'random'`` to
        # ``config['algorithms'] == {'random': {}}``.
        config['algorithms'] = algorithms.configuration
        return algorithms
```

```python
    def _is_different_from(self, config):
        is_diff = False
        for section, value in config.items():
            item = getattr(self, section)
            # See below in the comparison, that no special case is being considered for
            # ``section == 'algorithms'``. Because of this, ``is_new is True`` will be returned
            # falsely, if the previous steps in ``_sanitize_and_instantiate_config`` do not happen.
            # Special treatment of ``section == 'algorithms'``, will effectively do twice the thing that
            # `BaseAlgorithm.__init__` does through `PrimaryAlgo`.
            # It breaks delegation of responsibility.
            if item != value:
                is_diff = True
                break
```

What is being proposed above, does not resolve this case. The reason is that ``config['algorithms']`` is being replaced by an actual `PrimaryAlgo` instance which makes this key not immediately comparable with what (possibly) already exists in the database.

I am not very fond of this particular part of current implementation. But I think it is nonetheless a sufficient measure.

Reply:
OK! There was an important detail I missed. The method `_sanitize_and_instantiate_config()` will 
1) instantiate the space and algorithm(s) and will set the algorithm's config inside the experiment's config (**sanitize**), 
2) but it will return the instantiated algorithm, not its config (**instantiate**)

This is because 
1) We want a unique form of config to compare (ex: algo: random and algo: {random: {}} maps to the same config {random: {}}) (**sanitize**)
2) We want to instantiate the actual algorithm to set it as an attribute of the experiment (**instantiate**)

The problem is we need the instantiated config in order to get the sanitized config, that's why both operation are currently coupled in a single method.

I think we could get rid of the _sanitize_ part by doing the tests for differences and forking on a temporary copy of the experiment. This way we would still avoid changes on the experiment itself during the tests. Something like this:

```python
def configure(self, config):
    experiment = copy.deepcopy(self)
    experiment._instantiate_config(config)

    if self.status is None:
        is_new = True
    else:
        is_new = self._is_different_from(experiment.configuration)
        if is_new:
            experiment._fork_config(config)

    self._instantiate_config(experiment.configuration)

    self._init_done = True
    self.status = 'Pending'

    # ...
```

The method `_instantiate_config` would setup the experiment with instantiated objects of the config.

```python
def _instantiate_config(self, config):
    space = SpaceBuilder().build_from(config['metadata']['user_args'])
    if not space:
        raise ValueError("Parameter space is empty. There is nothing to optimize.")

    # Instantiate algorithms
    self.algorithms = PrimaryAlgo(space, config['algorithms'])

    for section,  value in config.items():
        setattr(self, section, value)
```

Reply:
Nice idea! Much cleaner!!